src/mscorlib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -349,22 +378,36 @@ public virtual void GetObjectData(SerializationInfo info, StreamingContext conte
             }
         }
 
-        private int FindEntry(TKey key)
+        private ref Entry FindEntry(TKey key, out bool found)
         {
-            if (key == null)
-            {
-                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.key);
-            }
+            Debug.Assert(key != null);
 
+            found = true;
             if (buckets != null)
             {
-                int hashCode = comparer.GetHashCode(key) & 0x7FFFFFFF;
-                for (int i = buckets[hashCode % buckets.Length]; i >= 0; i = entries[i].next)
+                int hashCode = (comparer == null ? EqualityComparer<TKey>.Default.GetHashCode(key) : comparer.GetHashCode(key)) & 0x7FFFFFFF;
```

Comment:
You can call `GetHashCode` directly on the key as enums don't box anymore.

Reply:
And I would extract it to separate methods. Though it  creates one more copy for complex keys but inlining of `EqualityComparer<TKey>.Default` methods [introduce a couple of copies](https://github.com/dotnet/coreclr/issues/1133) already.

```C#
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private int GetKeyHashCode(TKey key)
        {
            if (default(TKey) != null)
            {
                return comparer != null ? comparer.GetHashCode(key) : key.GetHashCode();
            }
            else
            {
                return comparer.GetHashCode(key);
            }
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private bool KeysEqual(TKey x, TKey y)
        {
            if (default(TKey) != null)
            {
                return comparer != null ? comparer.Equals(x, y) : EqualityComparer<TKey>.Default.Equals(x, y);
            }
            else
            {
                return comparer.Equals(x, y);
            }
        }
```

Reply:
Be careful about the class path ... AggressiveInlining is not always respected for it because of the runtime does not know how to inline the complex generic dictionary access that it introduces.

Reply:
Hmm.. it works for `int`as the key but I haven't yet checked tuples, complex structs and reference types.

Reply:
Naive branch prediction (non-loop) prefers continue than jump forward; so using default comparer in the first path (being more common than custom comparer) would be better.

Will change hashcode.

Reply:
That  exactly why I changed the order :)
```ASM
       ; comparer != null
       488B4E18             mov      rcx, gword ptr [rsi+24]
       4885C9               test     rcx, rcx
       750C                 jne      SHORT G_M33318_IG06
       ; inlined Defaul.Equals start
       3BD7                 cmp      edx, edi
       410F94C7             sete     r15b
       450FB6FF             movzx    r15, r15b 
       ; inlined Defaul.Equals end
       EB15                 jmp      SHORT G_M33318_IG07

G_M33318_IG06:
       448BC7               mov      r8d, edi
       49BB0001C36DFE070000 mov      r11, 0x7FE6DC30100
       3909                 cmp      dword ptr [rcx], ecx
       41FF13               call     qword ptr [r11]System.Collections.Generic.IEqualityComparer`1[Int32][System.Int32]:Equals(int,int):bool:this
       448BF8               mov      r15d, eax

G_M33318_IG07:
       4585FF               test     r15d, r15d
       751A                 jne      SHORT G_M33318_IG11
```

Reply:
Oh my days, you are right, the compilation reverses the tests 😢

Reply:
And yes, it doesn't inline for `Dictionary`2[__Canon, Int32]`  :cry: 
```
call     System.Collections.Generic.Dictionary`2[__Canon,Int32][System.__Canon,System.Int32]:KeysEqual(ref,ref):bool:this
```